Utilize Aggregator and CounterCache

cyx opened the pull request:
Closes #11 

I'm running an experiment now, and it seems this change doesn't
actually change the metric names visibly, and the change is more
behind the scenes.

## Counters

This is a relatively obvious change, and there's little to no
downside just doing this.

```ruby
10.times do |i|
  Pliny::Metrics.count "foo", i
end
```

Results in a single submission within a tick of:

```
{name: "foo", value: 45}
```

## Measurements

```ruby
10.times do |i|
  Pliny::Metrics.measure "foo" do
    sleep i
  end
end
```

That would submit a single measurement as follows:

```
{name: "foo", count: 10, sum: 45, min: 0, max: 9}
```

And that would all be accessible from your metric page by
choosing the appropriate summary function.

Contentious points:

- Should we just copy `CounterCache` from librato-rack, or do
  we add it as a dependency. I do not have a strong opinion
  on this one, having done more go recently, the "copy what
  you only need" approach is more frequently favored.
- Should we try to still use Queue as before? I'm leaning heavily
  to just stop doing it, as we are just unnecessarily bumping
  our rate limit use without much benefit (e.g. we'd still be using
  SSA regardless, the net numbers would be the same, etc)

cyx committed:
Utilize Aggregator and CounterCache

I'm running an experiment now, and it seems this change doesn't
actually change the metric names visibly, and the change is more
behind the scenes, e.g.:

This is a relatively obvious change, and there's little to no
downside just doing this.

```ruby
10.times do |i|
  Pliny::Metrics.count "foo", i
end
```

Results in a single submission within a tick of:

```
{name: "foo", value: 45}
```

```ruby
10.times do |i|
  Pliny::Metrics.measure "foo" do
    sleep i
  end
end
```

That would submit a single measurement as follows:

```
{name: "foo", count: 10, sum: 45, min: 0, max: 9}
```

And that would all be accessible from your metric page by
choosing the appropriate summary function.

Contentious points:

- Should we just copy `CounterCache` from librato-rack, or do
  we add it as a dependency. I do not have a strong opinion
  on this one, having done more go recently, the "copy what
  you only need" approach is more frequently favored.
- Should we try to still use Queue as before? I'm leaning heavily
  to just stop doing it, as we are just unnecessarily bumping
  our rate limit use without much benefit (e.g. we'd still be using
  SSA regardless, the net numbers would be the same, etc)

cyx commented:
One thing to note, I see that most of these apps are configured to do:

`Pliny::Metrics.backends << ...`

I believe we should also stop doing this, and start doing:

`Pliny::Metrics.backends = [...]`

Having us keep logging those metric lines to `logplex => splunk` adds some overhead -- which is ok if they're valid but if they're dev/null we should probably just configure appropriately. my 2c anyway :)

jmervine commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -1,49 +1,48 @@
           @interval      = interval
           @mutex         = Mutex.new
-          @metrics_queue = Queue.new
+          @counter_cache = ::Librato::Collector::CounterCache.new(default_tags: nil)
```
I'm guessing this is an in-memory cache? What happens if the thread dies, do we lose all cached metrics?

cyx replied:
We lose cached metrics! -- which was the same for the prior case anyway FWIW.

In the event you get graceful signaling via `#stop` you'd get the flush behavior as a best effort (previous and new version)

jmervine commented:
:+1:

appleton commented:
Looks good to me, @gudmundur could you check this too and then I'll ship an update. 👍

mikehale commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -1,49 +1,48 @@
         private
 
-        attr_reader :interval, :timer, :counter, :metrics_queue, :librato_queue
+        attr_reader :interval, :timer, :counter, :counter_cache, :aggregator, :librato_queue
```
`:counter` can be removed from this list

mikehale commented:
Left one comment, but otherwise, LGTM. Thanks for making this happen @cyx

mikehale commented:
While trying this branch on an app I ran into this error a few times:

```
Librato::Metrics::ClientError: {:status=>400, :url=>"https://metrics-api.librato.com/v1/metrics", :user_agent=>"librato-metrics/2.1.0 (ruby; 2.3.3p222; x86_64-linux) direct-faraday/0.11.0", :request_body=>"{\"gauges\":[],\"source\":\"high-volume-app.production\"}", :response_headers=>{"date"=>"Wed, 15 Mar 2017 18:03:54 GMT", "content-type"=>"application/json", "content-length"=>"90", "connection"=>"close", "server"=>"nginx"}, :response_body=>"{\"errors\":{\"params\":{\"measurements\":[\"No measurements found\"]}},\"request_time\":1489601034}"}
```

I'm not sure how I ended up with no metrics, since this app usually produces quite a few.

mikehale commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -54,21 +53,13 @@ def start_timer
+          sync do
+            counter_cache.flush_to(librato_queue)
+            librato_queue.merge!(aggregator)
+            librato_queue.submit
```
I think this should be changed to `librato_queue.submit unless librato_queue.empty?`. That should resolve the issue I mentioned [here](https://github.com/heroku/pliny-librato/pull/12#issuecomment-286833920).

mikehale replied:
Hmm, I tried this change and I'm still hitting the error. I'll dig around a bit and see if I can figure out what is happening.

cyx replied:
Yes the `#submit` in itself should already handle that case:

```ruby
      def submit
        return true if self.queued.empty?
        options = {per_request: @per_request}
        if persister.persist(self.client, self.queued, options)
          @last_submit_time = Time.now
          clear and return true
        end
        false
      rescue ClientError
        # clean up if we hit exceptions if asked to
        clear if @clear_on_failure
        raise
      end
```

mikehale commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -54,21 +53,13 @@ def start_timer
-          sync { librato_queue.submit }
+          sync do
+            counter_cache.flush_to(librato_queue)
+            librato_queue.merge!(aggregator)
```
It looks like `merge!` also submits. I think this should be `librato_queue.merge!(aggregator) unless aggregator.empty?`

mikehale commented:
I noticed that the tests are failing. I started re-working things in a branch to use Webmock to test at the HTTP interface instead of mocking out the internals of `Backend`. However, I'm at EOD and don't have things in a committable state yet, and I'm not sure if I will like the webmock version or not.

cyx committed:
Make measure thread safe

cyx committed:
Remove unused attribute `counter`

cyx committed:
Lock the counter cache when adding

cyx committed:
Don't submit twice even though second is noop

cyx committed:
Need to submit explicitly with low count

cyx committed:
Move submission to mutex

cyx committed:
Log it

cyx committed:
Ugh only submit if it's not empty?

mikehale commented:
Regarding appending to `backends` the readme currently recommends this:

```
# Add the following to config/initializers/librato.rb
Librato::Metrics.authenticate(Config.librato_email, Config.librato_key)
librato_backend = Pliny::Librato::Metrics::Backend.new(source: "myapp.production")
librato_backend.start
Pliny::Metrics.backends << librato_backend
```

This has the added problem that `librato.rb` is loaded before `metrics.rb` which does the following:

```
Pliny::Metrics.backends = [Pliny::Metrics::Backends::Logger]
```

... which effectively removes the librato metrics backend. Instead perhaps the readme should recommend updating metrics.rb:

```
# Change config/initializers/metrics.rb to something like this:

if Config.librato_user && Config.librato_token
  Librato::Metrics.authenticate(Config.librato_user, Config.librato_token)
  Pliny::Metrics.backends = [Pliny::Librato::Metrics::Backend.new(source: "#{Config.app_name}.#{Config.app_env}", interval: 30).tap(&:start)]
else
  Pliny::Metrics.backends = [Pliny::Metrics::Backends::Logger]
end
```

cyx commented:
:+1: on readme recommendations

joshwlewis commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -54,23 +54,20 @@ def start_timer
-              msg == POISON_PILL ? break : enqueue_librato(msg)
+        def flush_librato
+          sync do
+            $stderr.puts "pliny-librato: counter_cache.flush_to: %s" % counter_cache.instance_variable_get(:@cache).inspect
```
Can we send this to `Pliny.log` instead of `$stderr`?

cyx replied:
Hi sorry, we're actually working through a PR in librato-metrics since @mikehale found a bug there. Once that's landed, I can streamline this back again to its healthy form without logging.

gudmundur replied:
I'm assuming the change is this https://github.com/librato/librato-metrics/pull/137?

gudmundur commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -54,23 +54,20 @@ def start_timer
-              msg = metrics_queue.pop
-              msg == POISON_PILL ? break : enqueue_librato(msg)
+        def flush_librato
+          sync do
```
Any time `flush_librato` is running the `sync` here it has the side effect that `report_counts` and `report_measure` will block. The way we set this code up before was that reporting metrics was a non blocking and thread safe operation.

mikehale replied:
@cyx now that https://github.com/librato/librato-metrics/pull/137 has landed maybe we can try backing off on some of these sync blocks?

cyx committed:
Remove the submission from the mutex

cyx committed:
Do skip_measurement_times: true

This avoids having a `measure_time` field which we don't
need for counters.

cyx committed:
Update all specs

cyx committed:
Cleanup spec names

cyx committed:
Test that librato_queue.submit is not synchronized

cyx committed:
Update README

mikehale commented on README.md:
```diff
@@ -27,10 +27,15 @@ Or install it yourself as:
 Add a new initializer `config/initializers/librato.rb`:
```
I think this should also change from 'Add a new initializer' to 'Update the `config/initializers/metrics.rb`'. Otherwise because 'M' comes after 'L' the metrics initializer will overwrite the librato initializer.

joshwlewis committed:
Stop recommending a new initializer

The previous recommendation was a potential source for conflicts due to load order.

joshwlewis committed:
Drop the README bit about auto-submit counts

It's no longer a feature.

mikehale commented:
🍰

gudmundur commented on lib/pliny/librato/metrics/backend.rb:
```diff
@@ -54,21 +57,13 @@ def start_timer
+            librato_queue.merge!(aggregator)
+            aggregator.clear
+          end
+          librato_queue.submit
```
@cyx @joshwlewis The `Librato::Queue` is not thread safe and we have an interleaving now where this line can be called twice. When `#stop` is invoked and the timer loop calls `#flush_librato`.

Also as @mikehale pointed out above `Librato::Queue#merge!` will submit if the metric count threshold is hit.

I think all of these issues will need to be addressed as we're bound to run into them given enough time.

mikehale replied:
@gudmundur thanks for pointing this out... I'm assuming the "poison pill" approach that was used previously is how you avoided race conditions and blocking?

gudmundur replied:
We used a thread safe queue to bridge between receiving metrics and submitting them. If memory serves the poison pill was to drain the queue until the pill, then stop processing.

joshwlewis commented:
I think we could initialize a new Librato::Queue for each `flush_librato`
call, and we'd eliminate the dual access condition.
On Thu, Mar 23, 2017 at 8:21 AM Michael Hale <notifications@github.com>
wrote:

> *@mikehale* commented on this pull request.
> ------------------------------
>
> In lib/pliny/librato/metrics/backend.rb
> <https://github.com/heroku/pliny-librato/pull/12#discussion_r107665134>:
>
> >          def flush_librato
> -          sync { librato_queue.submit }
> +          sync do
> +            counter_cache.flush_to(librato_queue)
> +            librato_queue.merge!(aggregator)
> +            aggregator.clear
> +          end
> +          librato_queue.submit
>
> @gudmundur <https://github.com/gudmundur> thanks for pointing this out...
> I'm assuming the "poison pill" approach that was used previously is how you
> avoided race conditions and blocking?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/heroku/pliny-librato/pull/12#discussion_r107665134>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4AqBoBF-dRcv30yBieiFD75Uie0s74ks5ronHEgaJpZM4MZtgJ>
> .
>

cyx commented:
@joshwlewis exactly what I have right now! I'm about to push the change.
